rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/util/SemanticTokenizer.java

```diff
@@ -514,4 +538,71 @@ else if (Character.isSupplementaryCodePoint(currentChar)) {
             }
         }
     }
+
+    // The idea behind the patch is to dynamically map productions in Rascal's
+    // own grammar with no category (e.g, number literals), or with a legacy
+    // category (e.g., string literals), to current categories (i.e., semantic
+    // token types).
+    private static class RascalPatch {
+
+        // As an optimization, productions that have already been mapped in the
+        // past, are cached in an identity map/set for fast lookups in the
+        // future.
+        Map<IConstructor, String> positiveCache = new IdentityHashMap<>();
+        Set<IConstructor> negativeCache = Collections.newSetFromMap(new IdentityHashMap<>());
+
+        @SuppressWarnings("java:S131") // Switches without defaults are intended in this method
+        public String apply(IConstructor prod, String defaultCategory) {
+
+            // Check the caches
+            if (positiveCache.containsKey(prod)) {
+                return positiveCache.get(prod);
+            }
+            if (negativeCache.contains(prod)) {
+                return defaultCategory; // Possibly different each time the negative cache is hit
+            }
+
+            // Apply the patch (positively)
+            var def = ProductionAdapter.getDefined(prod);
+
+            if (isLabeledLiteral(def)) {
+                switch (SymbolAdapter.getLabel(def)) {
+                case "integer":
+                case "real":
+                case "rational":
+                    return putIntoPositiveCache(prod, SemanticTokenTypes.Number);
+                case "location":
+                    return putIntoPositiveCache(prod, SemanticTokenTypes.String);
+                case "regExp":
+                    return putIntoPositiveCache(prod, SemanticTokenTypes.Regexp);
+                }
+            }
+
+            if (SymbolAdapter.isLex(def)) {
```

Comment:
is this not an else if?

Reply:
It could have been, yes, but it's gone now anyway (next comment)